src/stan/io/deserializer.hpp

```diff
@@ -0,0 +1,1978 @@
+   *
+   * <p>See <code>stan::math::corr_constrain(T,T&)</code>.
+   *
+   * @tparam Ret The type to return.
+   * @tparam Jacobian Whether to increment the log of the absolute Jacobian
+   * determinant of the transform.
+   * @tparam LP Type of log probability.
+   * @param k Dimensions of matrix return type.
+   * @param lp The reference to the variable holding the log
+   * probability to increment.
+   */
+  template <typename Ret, bool Jacobian, typename LP,
+            require_not_std_vector_t<Ret>* = nullptr,
+            require_matrix_t<Ret>* = nullptr>
+  auto read_corr(LP& lp, size_t k) {
+    using stan::math::corr_matrix_constrain;
+    if (Jacobian) {
+      return corr_matrix_constrain(
+          this->read<conditional_var_val_t<Ret, vector_t>>((k * (k - 1)) / 2),
+          k, lp);
+    } else {
+      return corr_matrix_constrain(
+          this->read<conditional_var_val_t<Ret, vector_t>>((k * (k - 1)) / 2),
+          k);
+    }
+  }
+
+  /**
+   * Specialization of `read_corr` for `std::vector` return types.
+   *
+   * <p>See <code>stan::math::corr_constrain(T,T&)</code>.
+   *
+   * @tparam Ret The type to return.
+   * @tparam Jacobian Whether to increment the log of the absolute Jacobian
+   * determinant of the transform.
+   * @tparam LP Type of log probability.
+   * @tparam Sizes A parameter pack of integral types.
+   * @param lp The reference to the variable holding the log
+   * @param vecsize The size of the return vector.
+   * @param sizes Pack of integrals to use to construct the return's type.
+   * probability to increment.
+   * @return The next scalar transformed to a correlation.
+   */
+  template <typename Ret, bool Jacobian, typename LP, typename... Sizes,
+            require_std_vector_t<Ret>* = nullptr>
+  auto read_corr(LP& lp, const size_t vecsize, Sizes... sizes) {
+    std::decay_t<Ret> ret;
+    ret.reserve(vecsize);
+    for (size_t i = 0; i < vecsize; ++i) {
+      ret.emplace_back(
+          this->read_corr<value_type_t<Ret>, Jacobian>(lp, sizes...));
+    }
+    return ret;
+  }
+
+  /**
+   * Return a unit_vector of the specified size made up of the
+   * next scalars.
+   *
+   * <p>See <code>stan::math::check_unit_vector</code>.
+   *
+   * @tparam Ret The type to return.
+   * @param k Size of returned unit_vector
+   * @return unit_vector read from the specified size number of scalars
+   * @throw std::runtime_error if the next k values is not a unit_vector
+   * @throw std::invalid_argument if k is zero
+   */
+  template <typename Ret, require_vector_t<Ret>* = nullptr>
+  auto read_unit_vector(size_t k) {
+    if (unlikely(k == 0)) {
+      []() STAN_COLD_PATH {
+        std::string msg = "deserializer: unit vectors cannot be size 0.";
+        throw std::invalid_argument(msg);
+      }();
+    }
+    auto ret = read<Ret>(k);
+    using stan::math::check_unit_vector;
+    using stan::math::value_of;
+    check_unit_vector("deserializer", "Unit Vector", value_of(ret));
```

Comment:
`check_unit_vector` has a non-zero size check in it ([here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/err/check_unit_vector.hpp)). Since `read<Ret>(0)` is okay, we don't need the check before the read.